bindings/go/oci/repository.go

```diff
@@ -0,0 +1,690 @@
+	// - the Source Access from descriptor.Resource
+	// - the Target Access from the given target specification
+	// It might be that during the upload, the source pointer may be updated with information gathered during upload
+	// (e.g. digest, size, etc).
+	//
+	// The content of form blob.ReadOnlyBlob is expected to be a (optionally gzipped) tar archive that can be read with
+	// oci.NewFromTar, which interprets the blob as an OCILayout.
+	//
+	// The given OCI Layout MUST contain the resource described in source with an v1.OCIImage specification,
+	// otherwise the upload will fail
+	UploadResource(ctx context.Context, targetAccess runtime.Typed, source *descriptor.Resource, content blob.ReadOnlyBlob) (err error)
+
+	// DownloadResource downloads a resource from the repository.
+	// THe resource MUST contain a valid v1.OCIImage specification that exists in the Store.
+	// Otherwise, the download will fail.
+	//
+	// The blob.ReadOnlyBlob returned will always be an OCI Layout, readable by oci.NewFromTar.
+	// For more information on the download procedure, see NewOCILayoutWriter.
+	DownloadResource(ctx context.Context, res *descriptor.Resource) (content blob.ReadOnlyBlob, err error)
+}
+
+// Resolver defines the interface for resolving references to OCI stores.
+type Resolver interface {
+	// StoreForReference resolves a reference to a Store.
+	// Each component version can resolve to a different store.
+	StoreForReference(ctx context.Context, reference string) (Store, error)
+
+	// ComponentVersionReference returns a unique reference for a component version.
+	ComponentVersionReference(component, version string) string
+}
+
+// Store defines the interface for interacting with an OCI store.
+type Store interface {
+	content.ReadOnlyStorage
+	content.Pusher
+	content.TagResolver
+	content.Tagger
+}
+
+// Repository implements the ComponentVersionRepository interface using OCI registries.
+// Each component version is stored in a separate OCI repository.
+type Repository struct {
+	scheme *runtime.Scheme
+
+	// localBlobMemory temporarily stores local blobs until they are added to a component version.
+	// thus any local blob added with AddLocalResource will be added to the memory until
+	// AddComponentVersion is called with a reference to that resource.
+	// Note that Store implementations are expected to either allow orphaned LocalBlobs or
+	// regularly issue an async garbage collection to remove them.
+	localBlobMemory LocalBlobMemory
+
+	// resolver resolves component version references to OCI stores.
+	resolver Resolver
+}
+
+// RepositoryFromResolverAndMemory creates a new Repository instance.
+// This is a convenience function that uses the new options pattern.
+func RepositoryFromResolverAndMemory(resolver Resolver, memory LocalBlobMemory) *Repository {
+	repo, err := NewRepository(
+		WithResolver(resolver),
+		WithLocalBlobMemory(memory),
+	)
+	if err != nil {
+		panic(err)
```

Comment:
`panic`?

Reply:
I mean, shouldn't this be a return error?

Reply:
removed